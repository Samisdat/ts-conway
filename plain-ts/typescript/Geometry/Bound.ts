export class Bound {

    private min: number;

    private max: number;

    constructor(min: number, max: number) {

        if (max < min) {
            throw new Error('max can not be lower then min');
        }

        this.min = min;
        this.max = max;
    }

    public isWithin(test: number): boolean {
        return (test >= this.min && test <= this.max);
    }

    public isBelow(test: number): boolean {
        return (test < this.min);
    }

    public isAbove(test: number): boolean {
        return (test > this.max);
    }

    public confine(value: number): number {

        if (value < this.min) {
            return this.min;
        }

        if (value > this.max) {
            return this.max;
        }

        return value;

    }

    public getMax(): number{
        return this.max;
    }

    public getMin(): number{
        return this.min;
    }

    public expand(value: number): void{

        if(true === this.isWithin(value)){
            return;
        }

        if(true === this.isBelow(value)){
            this.min = value
        }

        if(true === this.isAbove(value)){
            this.max = value
        }

    }

}
