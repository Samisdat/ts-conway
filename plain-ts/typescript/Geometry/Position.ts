export class Position {

    public readonly  x: number;
    public readonly  y: number;

    constructor(x: number, y: number) {

        if (-0 === x) {
            x = 0;
        }

        if (-0 === y) {
            y = 0;
        }

        this.x = x;
        this.y = y;
    }

    public move(position: Position): Position {
        return new Position(
            this.x + position.x,
            this.y + position.y
        );
    }

    public inverse(): Position {
        return new Position(
            -1 * this.x,
            -1 * this.y
        );
    }

    public compare(position: Position): boolean {

        return (position.x === this.x && position.y === this.y);

    }

    public getNeighbours(): Position[] {

        const neighbours: Position[] = [];

        // neighbours on bottom
        neighbours.push(
            new Position(this.x - 1, this.y - 1)
        );
        neighbours.push(
            new Position(this.x, this.y - 1)
        );
        neighbours.push(
            new Position(this.x + 1, this.y - 1)
        );

        // neighbours on the left and on the right
        neighbours.push(
            new Position(this.x - 1, this.y)
        );
        neighbours.push(
            new Position(this.x + 1, this.y)
        );

        // neighbours on top
        neighbours.push(
            new Position(this.x - 1, this.y + 1)
        );
        neighbours.push(
            new Position(this.x, this.y + 1)
        );
        neighbours.push(
            new Position(this.x + 1, this.y + 1)
        );

        return neighbours;
    }

    public clone(): Position {
        return new Position(
            this.x,
            this.y
        );
    }

    public toString(): string {

        let str = '';

        str += (0 <= this.x) ? '+' : '';
        str += this.x;

        str += '/';

        str += (0 <= this.y) ? '+' : '';
        str += this.y;

        return str;

    }
}
