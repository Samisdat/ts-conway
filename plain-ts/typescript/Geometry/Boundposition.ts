import {Position} from '@Conway/Geometry/Position';
import {Bound} from '@Conway/Geometry/Bound';

export class Boundposition {

    private x: Bound;
    private y: Bound;

    constructor(topLeft: Position, bottomRight: Position) {

        if (bottomRight.x < topLeft.x) {
            throw new Error('bottomRight must be right from topLeft');
        }

        if (bottomRight.y < topLeft.y) {
            throw new Error('bottomRight must be below from topLeft');
        }

        this.x = new Bound(topLeft.x, bottomRight.x);
        this.y = new Bound(topLeft.y, bottomRight.y);

    }

    public isWithin(test: Position): boolean {
        return (this.x.isWithin(test.x) && this.y.isWithin(test.y));
    }

    public confine(value: Position): Position {

        return new Position(
            this.x.confine(value.x),
            this.y.confine(value.y)
        );

    }

    public bottomLeft(): Position{

        return new Position(this.x.getMin(), this.y.getMin());

    }

    public topRight(): Position{

        return new Position(this.x.getMax(), this.y.getMax());

    }

    public getWidth(): number{

        return this.x.getMax() - this.x.getMin() + 1;

    }

    public getHeight(): number{

        return this.y.getMax() - this.y.getMin() + 1;

    }

    public expand(position: Position): void{

        if(true === this.isWithin(position)){
            return;
        }

        this.x.expand(position.x);
        this.y.expand(position.y);

    }

}
