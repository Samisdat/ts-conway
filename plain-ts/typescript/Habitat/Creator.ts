import {Position} from '@Conway/Geometry/Position';
import {Dimension} from '@Conway/Habitat/Dimension';

export class Creator {

    public readonly canvasWidth: number;
    public readonly canvasHeigth: number;
    public readonly originalCellWidthAndHeight: number;
    public readonly cellWidthAndHeight: number;
    public readonly pan: Position;

    public readonly sourcePosition: Position;
    public readonly offset: Position;

    public readonly zoom: number;

    private dimension: Dimension;

    constructor(
        canvasWidth: number,
        canvasHeight: number,
        cellWidthAndHeight: number,
        pan: Position,
        zoom: number
    ) {
        this.canvasWidth = canvasWidth;
        this.canvasHeigth = canvasHeight;
        this.originalCellWidthAndHeight = cellWidthAndHeight;
        this.pan = pan;
        this.zoom = zoom;

        this.offset = new Position(
            this.pan.x % 1,
            this.pan.y % 1,

        );

        const sourcePosition = this.pan.clone().move(this.offset.inverse());
        this.sourcePosition = new Position(
            sourcePosition.x,
            sourcePosition.y
        );

        this.cellWidthAndHeight = this.originalCellWidthAndHeight * this.getZoom();

        this.setDimension();

    }

    private setDimension(): void{

        let rows = Math.ceil(this.canvasWidth / this.cellWidthAndHeight);

        if (0 === rows % 2) {
            rows += 1;
        }

        if (0 !== this.offset.x) {
            rows += 2;
        }

        let cols = Math.ceil(this.canvasHeigth / this.cellWidthAndHeight);

        if (0 === cols % 2) {
            cols += 1;
        }

        if (0 !== this.offset.y) {
            cols += 2;
        }

        this.dimension = new Dimension(rows, cols);

    }

    public getRows(): number {
        return this.dimension.getRows();
    }

    public getCols(): number {
        return this.dimension.getCols();
    }

    public getPan(): Position {
        return this.pan;
    }

    public getSourcePosition(): Position {
        return this.sourcePosition;
    }

    public getOffset(): Position {
        return this.offset.inverse();
    }

    public getZoom(): number {
        return this.zoom;
    }

    public getDimension(): Dimension {
        return this.dimension;
    }

}
