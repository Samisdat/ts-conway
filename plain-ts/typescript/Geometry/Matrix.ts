import {Position} from '@Conway/Geometry/Position';
import {Boundposition} from '@Conway/Geometry/Boundposition';

interface MatrixInterface {
    [index: string]: Position;
}


export class Matrix {

    private matrix: MatrixInterface = {};

    private boundposition: Boundposition;

    public has(position: Position): boolean {

        return (undefined !== this.matrix[position.toString()]);

    }

    public add(position: Position): void {

        if (true === this.has(position)) {
            return;
        }

        this.matrix[position.toString()] = position;


        if(undefined === this.boundposition){
            this.boundposition = new Boundposition(position, position);
        }
        else{
            this.boundposition.expand(position);
        }

    }

    public remove(position: Position): void {

        if (false === this.has(position)) {
            return;
        }

        delete this.matrix[position.toString()];

    }

    public getAll(): Position[] {

        const positions: Position[] = [];

        for (const index in this.matrix) {

            positions.push(this.matrix[index]);

        }

        return positions;

    }

    public getBound(): Boundposition {

        return this.boundposition;

    }

    public getWidth(): number {

        return this.getBound().getWidth();

    }

    public getHeight(): number {

        return this.getBound().getHeight();

    }

}
