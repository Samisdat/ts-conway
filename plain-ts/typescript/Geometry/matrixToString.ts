import {Matrix} from '@Conway/Geometry/Matrix';
import {Position} from '@Conway/Geometry/Position';

interface NumberedScale {
    sign: string[];
    number: string[];
};

const getSigns = (numberedScale: NumberedScale, i: number) =>{
    if(0 === i % 5){

        numberedScale.number.push(Math.abs(i).toString());

        if(0 > i){
            numberedScale.sign.push('-');
        }
        else if(0 < i){
            numberedScale.sign.push('+');
        }
        else{
            numberedScale.sign.push(' ');
        }

    }
    else{
        numberedScale.number.push(' ');
        numberedScale.sign.push(' ');
    }

};

const getScaleHorizontal = (matrix: Matrix, matrixArray: string[][]): string[][] => {

    const left = matrix.getBound().bottomLeft().x;
    const right  = matrix.getBound().topRight().x;

    matrixArray = matrixArray.reverse();

    const newRow: string[] = []
    for(let i = 0; i < matrixArray[0].length; i += 1){
        newRow.push('')
    }

    matrixArray.push(newRow);

    const scale: NumberedScale = {
        sign: [],
        number: [],
    };

    for(let i = left; i <= right; i += 1){

        getSigns(scale, i);
    }

    matrixArray.push(scale.number);
    matrixArray.push(scale.sign);

    matrixArray = matrixArray.reverse();
    return matrixArray;


};

const getScaleVertical = (matrix: Matrix, matrixArray: string[][]): string[][] => {

    const top  = matrix.getBound().topRight().y;
    const bottom = matrix.getBound().bottomLeft().y;

    const scale: NumberedScale = {
        sign: [' ', ' ', ' '],
        number: [' ', ' ', ' '],
    };

    for(let i = top; i >= bottom; i -= 1){

        getSigns(scale, i);

    }

    for(let i = 0; i < matrixArray.length; i += 1){

        matrixArray[i] = matrixArray[i].reverse();

        matrixArray[i].push(' ')
        matrixArray[i].push(' ')

        matrixArray[i].push(scale.number[i])
        matrixArray[i].push(scale.sign[i])

        matrixArray[i] = matrixArray[i].reverse();

    }

    return matrixArray;

};

const addScale = (matrix: Matrix, matrixString: string[][]): string[][] => {

    matrixString = getScaleHorizontal(matrix, matrixString);
    matrixString = getScaleVertical(matrix, matrixString);
    return matrixString;
};

export const matrixToString = (matrix: Matrix, scale = true): string => {

    if(0 === matrix.getAll().length){
        return 'Matrix is empty';
    }

    const width = matrix.getWidth();
    const height = matrix.getHeight();

    const topLeft = new Position(
        matrix.getBound().bottomLeft().x,
        matrix.getBound().topRight().y
    );

    let matrixOfStrings: string[][] = [];

    let pointer = topLeft.clone();

    for(let y = 0; y < height; y += 1){

        const line = [];

        for(let x = 0; x < width; x += 1){

            if(true === matrix.has(pointer)){
                line.push('O');
            }
            else{
                line.push('.');
            }

            pointer = pointer.move(new Position(1,0));

        }

        matrixOfStrings.push(line);

        pointer = new Position(
            topLeft.x,
            pointer.y - 1
        );

    }

    if(true === scale){
        matrixOfStrings = addScale(matrix, matrixOfStrings);
    }

    return matrixOfStrings.map((row)=>{return row.join('')}).join('\n');;

}
