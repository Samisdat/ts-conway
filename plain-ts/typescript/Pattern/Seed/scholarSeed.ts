import {SeedFunction} from '@Conway/Frontend/Web';
import {Position} from '@Conway/Geometry/Position';
import {readPatternFromPlainFile} from '@Conway/Pattern/readPatternFromPlainFile';
import {scholar} from '@Conway/Pattern/Store/scholar.cells';
import {Creator} from '@Conway/Habitat/Creator.ts';
import {Matrix} from '@Conway/Geometry/Matrix';
import {seedPattern} from '@Conway/Pattern/seedPattern';

export const scholarSeed: SeedFunction = (gridCreator: Creator, matrix: Matrix) => {

    const pattern = readPatternFromPlainFile(scholar);

    pattern.rotate();

    seedPattern(
        matrix,
        pattern,
        new Position(-75, 0)
    );

    seedPattern(
        matrix,
        pattern,
        new Position(75, 0)
    );

};

