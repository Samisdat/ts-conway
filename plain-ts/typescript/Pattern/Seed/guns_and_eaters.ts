import {SeedFunction} from '@Conway/Frontend/Web';
import {Position} from '@Conway/Geometry/Position';
import {readPatternFromPlainFile} from '@Conway/Pattern/readPatternFromPlainFile';
import {gunsAndEatersPattern} from '@Conway/Pattern/Store/gunsAndEaters.cells';
import {Creator} from '@Conway/Habitat/Creator.ts';
import {seedPattern} from '@Conway/Pattern/seedPattern';
import {Matrix} from '@Conway/Geometry/Matrix';
import {matrixToString} from '@Conway/Geometry/matrixToString';


export const gunsAndEaters: SeedFunction = (gridCreator: Creator, matrix: Matrix) => {

    const gunsAndEaters = readPatternFromPlainFile(gunsAndEatersPattern);

    const patternGutter = 2;

    let repeat = Math.floor(gridCreator.getRows() / (gunsAndEaters.getWidth() + patternGutter ));

    if (0 === repeat % 2) {
        repeat += 1;
    }

    const patternsPerSide = Math.floor(repeat / 2);

    seedPattern(matrix,
        gunsAndEaters,
        new Position(0, -1)
    );

    let move = new Position(0, -1);

    for (let i = 0; i < patternsPerSide; i += 1) {

        move = move.move(
            new Position(
                -1 * (gunsAndEaters.getWidth() + patternGutter), 0
            )
        );

        seedPattern(matrix,
            gunsAndEaters,
            move
        );

    }

    move = new Position(0, -1);

    for (let i = 0; i < patternsPerSide; i += 1) {

        move = move.move(
            new Position(
                gunsAndEaters.getWidth() + patternGutter, 0
            )
        );

        seedPattern(matrix,
            gunsAndEaters,
            move
        );

    }

};
