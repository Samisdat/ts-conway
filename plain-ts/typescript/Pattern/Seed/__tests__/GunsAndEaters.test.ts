import serializer from '../../../../jest-serialize-conway';
import {Position} from '@Conway/Geometry/Position';
import {gunsAndEaters} from '@Conway/Pattern/Seed/guns_and_eaters';
import {Creator} from '@Conway/Habitat/Creator.ts';
import {Matrix} from '@Conway/Geometry/Matrix';

describe('guns_and_eaters', () => {

    beforeAll(() => {

        expect.addSnapshotSerializer(serializer);

    });

    test('seed with even grid', () => {

        const gridCreator = new Creator(
            200,
            100,
            1,
            new Position(0, 0),
            1
        );

        const matrix = new Matrix();

        gunsAndEaters(gridCreator, matrix);

        expect(matrix).toMatchSnapshot();

    });

    test('seed with odd grid', () => {

        const gridCreator = new Creator(
            50,
            100,
            1,
            new Position(0, 0),
            1
        );

        const matrix = new Matrix();

        gunsAndEaters(gridCreator, matrix);

        expect(matrix).toMatchSnapshot();

    });

});
