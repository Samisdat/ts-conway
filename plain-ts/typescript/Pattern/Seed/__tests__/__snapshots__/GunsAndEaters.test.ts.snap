// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`guns_and_eaters seed with even grid 1`] = `
OOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOO
OOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOO
OOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXXOOOOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXXOOOOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXXOOOOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXXOOOOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXX
OOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXXOOOOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXXOOOOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXXOOOOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXXOOOOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXX
OOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOOOOOOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOOOOOOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOOOOOOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOOOOOOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOO
OOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOO
OOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOOOOOOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOOOOOOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOOOOOOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOOOOOOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOO
XXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOOOOXXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOOOOXXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOOOOXXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOOOOXXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOO
XXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOOOOXXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOOOOXXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOOOOXXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOOOOXXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOO
OOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOO
OOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOO
`;

exports[`guns_and_eaters seed with odd grid 1`] = `
OOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOO
OOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOO
OOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXX
OOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXX
OOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOO
OOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOO
OOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOO
XXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOO
XXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOO
OOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOO
OOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOO
`;
